src/debug/di/rsmain.cpp

```diff
@@ -1953,29 +1967,60 @@ void Cordb::EnsureCanLaunchOrAttach(BOOL fWin32DebuggingEnabled)
 
 HRESULT Cordb::CreateObjectV1(REFIID id, void **object)
 {
-    return CreateObject(CorDebugVersion_1_0, id, object);
+    return CreateObject(CorDebugVersion_1_0, ProcessDescriptor::UNINITIALIZED_PID, NULL, id, object);
 }
 
 #if defined(FEATURE_DBGIPC_TRANSPORT_DI)
 // CoreCLR activates debugger objects via direct COM rather than the shim (just like V1). For now we share the
 // same debug engine version as V2, though this may change in the future.
 HRESULT Cordb::CreateObjectTelesto(REFIID id, void ** pObject)
 {
-    return CreateObject(CorDebugVersion_2_0, id, pObject);
+    return CreateObject(CorDebugVersion_2_0, ProcessDescriptor::UNINITIALIZED_PID, NULL, id, pObject);
 }
 #endif // FEATURE_DBGIPC_TRANSPORT_DI
 
 // Static
 // Used to create an instance for a ClassFactory (thus an external ref).
-HRESULT Cordb::CreateObject(CorDebugInterfaceVersion iDebuggerVersion, REFIID id, void **object)
+HRESULT Cordb::CreateObject(CorDebugInterfaceVersion iDebuggerVersion, DWORD pid, LPCWSTR lpApplicationGroupId, REFIID id, void **object)
 {
     if (id != IID_IUnknown && id != IID_ICorDebug)
         return (E_NOINTERFACE);
 
-    Cordb *db = new (nothrow) Cordb(iDebuggerVersion);
+    LPSTR applicationGroupId = NULL;
+    if (lpApplicationGroupId != NULL)
+    {
+        // Get length of target string
+        int cbMultiByte = WideCharToMultiByte(CP_ACP, 0, lpApplicationGroupId, -1, NULL, 0, NULL, NULL);
+        if (cbMultiByte == 0)
+        {
+            return E_FAIL;
+        }
+
+        applicationGroupId = new (nothrow) CHAR[cbMultiByte];
+        if (applicationGroupId == NULL)
+        {
+            return (E_OUTOFMEMORY);
+        }
+
+        /* Convert to ASCII */
+        cbMultiByte = WideCharToMultiByte(CP_ACP, 0, lpApplicationGroupId, -1, applicationGroupId, cbMultiByte, NULL, NULL);
+        if (cbMultiByte == 0)
+        {
+            return E_FAIL;
```

Comment:
I've just noticed the applicationGroupId leaks here.

Reply:
It is cleaned up in the cordb destructor. 

```
Cordb::~Cordb()
{
    LOG((LF_CORDB, LL_INFO10, "C::~C Terminating Cordb object.\n"));
    if (m_pd.m_ApplicationGroupId != NULL)
    {
        delete [] m_pd.m_ApplicationGroupId;
    }
    g_pRSDebuggingInfo->m_Cordb = NULL;
}
```

Reply:
It is not - this is a local variable applicationGroupId , not the member. And it is not assigned to the member when we leave at this spot unless I am missing something.

Reply:
Unless I'm missing something, the "applicationGroupId" local is passed to ProcessDescriptor::Create which initializes and saves this pointer to the "m_ApplicationGroupId" field in the local "pd" which is passed to the Cordb constructor.  The constructor saves pd to m_pd and in the cordb destructor the m_pd.m_ApplicationGroupId field is deleted which is the original allocation in Cordb::CreateObject().

Reply:
But that doesn't happen if we exit at line 2009. We've allocated applicationGroupId local at line 1999 and we call the ProcessDescriptor::Create at line 2013

Reply:
Actually "NewArrayHolder" is what the runtime should use. Instead of "Detach" use "Extract" to get the pointer to pass to ProcessDescriptor::Create.

Reply:
@odhanson, never mind using a holder.  It gets tricky to deal with the Cordb out of memory error path.  Just explicitly delete it at line 2009.